src/i2c.cpp

```diff
@@ -108,522 +242,571 @@ void I2C::init(const i2c_hardware_struct_t *c)
+  {
+    log_line;
+  }
+
   //reset errors
-  I2C_ClearFlag(c_->dev, I2C_SR1_OVR | I2C_SR1_AF | I2C_SR1_ARLO | I2C_SR1_BERR);
-  current_status_ = IDLE;
-  log_line;
-  transfer_complete_cb();
-}
+  if (c_->dev->SR1 & AF)
+  {
+    log_line;
+    c_->dev->SR1 &= ~AF;
+  }
+  if (c_->dev->SR1 & BERR)
+  {
+    log_line;
+    c_->dev->SR1 &= ~BERR;
+  }
 
-// This is the I2C_IT_EV handler
-void I2C::handle_event()
-{
-  uint32_t last_event = I2C_GetLastEvent(c_->dev);
-  interrupt_history_.add_event(c_->dev->SR2 << 16 | c_->dev->SR1);
-  // We just sent a byte
-  if ((last_event & I2C_EVENT_MASTER_BYTE_TRANSMITTED) == I2C_EVENT_MASTER_BYTE_TRANSMITTED)
+  if (current_task_) // run the STOP task if there is one
   {
-    last_event_us_ = micros();
-    // If we are reading, then we just sent a subaddress and need to send
-    // a repeated start, and enable the DMA NACK
-    if (current_status_ == READING)
-    {
-      log_line;
-      I2C_AcknowledgeConfig(c_->dev, ENABLE);
-      I2C_DMALastTransferCmd(c_->dev, ENABLE);
-      I2C_GenerateSTART(c_->dev, ENABLE);
-    }
-    // We are in write mode and are done, need to clean up
-    else
-    {
-      log_line;
-      I2C_GenerateSTOP(c_->dev, ENABLE);
-      I2C_ITConfig(c_->dev, I2C_IT_EVT | I2C_IT_ERR, DISABLE);
-      transfer_complete_cb();
-    }
+    log_line;
+    handleTask();
   }
-  
-  // We just sent the address in write mode
-  else if ((last_event & I2C_EVENT_MASTER_TRANSMITTER_MODE_SELECTED) == I2C_EVENT_MASTER_TRANSMITTER_MODE_SELECTED)
+  else // otherwise add a STOP task and run it
   {
-    last_event_us_ = micros();
-    // We need to send the subaddress
-    if (!subaddress_sent_)
-    {
-      log_line;
-      I2C_SendData(c_->dev, reg_);
-      subaddress_sent_ = true;
-      if (current_status_ == WRITING)
-      {
-        log_line;
-        I2C_SendData(c_->dev, data_);
-        done_ = true;
-      }
-    }
-    // We need to send our data (no subaddress)
-    else
+    log_line;
+    beginJob();
+    addTaskStop();
+    if (finalizeJob())
     {
       log_line;
-      I2C_SendData(c_->dev, data_);
-      done_ = true;
+      current_task_ = task_queue_.current();
+      handleTask();
     }
   }
-  
-  // We are in receiving mode, preparing to receive the big DMA dump
-  else if ((last_event & I2C_EVENT_MASTER_RECEIVER_MODE_SELECTED) == I2C_EVENT_MASTER_RECEIVER_MODE_SELECTED)
+}
+
+void I2C::unstick()
+{
+  I2C_Cmd(c_->dev, DISABLE);
+
+  I2C_ClearFlag(c_->dev, I2C_FLAG_BUSY);
+
+  // Turn off the interrupts
+  I2C_ITConfig(c_->dev, I2C_IT_EVT | I2C_IT_ERR, DISABLE);
+
+  //reset errors
+  I2C_ClearFlag(c_->dev, I2C_SR1_OVR | I2C_SR1_AF | I2C_SR1_ARLO | I2C_SR1_BERR);
+
+  scl_.write(GPIO::HIGH);
+  sda_.write(GPIO::HIGH);
+  scl_.set_mode(GPIO::OUTPUT, GPIO::HIGH);
+  sda_.set_mode(GPIO::OUTPUT, GPIO::HIGH);
+
+  delayMicroseconds(10);
+
+  static const int cyc = 200;
+
+  // clock out some bits
+  for (int i = 0; i < 18; ++i)
```

Comment:
This is not something we want to be doing while flying.

What is left on `unstick`?